File: backend/app/models/range_report.py
"""RangeReport — LLM-generated narrative report for an arbitrary date range."""

from __future__ import annotations

import uuid
from datetime import date, datetime
from typing import TYPE_CHECKING, Any

from sqlalchemy import Date, DateTime, ForeignKey, Text, UniqueConstraint, func
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base

if TYPE_CHECKING:
    from app.models.user import User


class RangeReport(Base):
    __tablename__ = "range_reports"
    __table_args__ = (
        UniqueConstraint(
            "user_id", "period_start", "period_end", name="range_reports_user_period_unique"
        ),
    )

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
    )
    period_start: Mapped[date] = mapped_column(Date, nullable=False)
    period_end: Mapped[date] = mapped_column(Date, nullable=False)
    headline: Mapped[str] = mapped_column(Text, nullable=False)
    body_md: Mapped[str] = mapped_column(Text, nullable=False)
    takeaways: Mapped[list[dict[str, Any]]] = mapped_column(JSONB, nullable=False, default=list)
    generated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )

    user: Mapped[User] = relationship("User")

File: frontend/lib/features/insight/data/range_report.dart
class RangeTakeaway {
  final String title;
  final String body;
  const RangeTakeaway({required this.title, required this.body});

  factory RangeTakeaway.fromJson(Map<String, dynamic> json) =>
      RangeTakeaway(title: json['title'] as String, body: json['body'] as String);
}

class RangeReport {
  final DateTime periodStart;
  final DateTime periodEnd;
  final String headline;
  final String bodyMd;
  final List<RangeTakeaway> takeaways;
  final DateTime generatedAt;

  const RangeReport({
    required this.periodStart,
    required this.periodEnd,
    required this.headline,
    required this.bodyMd,
    required this.takeaways,
    required this.generatedAt,
  });

  factory RangeReport.fromJson(Map<String, dynamic> json) => RangeReport(
        periodStart: DateTime.parse(json['period_start'] as String),
        periodEnd: DateTime.parse(json['period_end'] as String),
        headline: json['headline'] as String,
        bodyMd: json['body_md'] as String,
        takeaways: (json['takeaways'] as List<dynamic>? ?? const [])
            .map((e) => RangeTakeaway.fromJson(e as Map<String, dynamic>))
            .toList(),
        generatedAt: DateTime.parse(json['generated_at'] as String),
      );
}
